devito/finite_differences/finite_difference.py

```diff
@@ -0,0 +1,401 @@
+    diff = dim.spacing
+    idx = list(set([(dim + int(i+.5+off)*diff)
+                    for i in range(-int(fd_order / 2), int(fd_order / 2))]))
+    if int(fd_order / 2) == 1:
+        idx = [dim - diff, dim]
+    c = finite_diff_weights(deriv_order, idx, dim + off*dim.spacing)[-1][-1]
+    deriv = 0
+    for i in range(0, len(idx)):
+            deriv += expr.subs({dim: idx[i]}) * c[i]
+
+    return deriv.evalf(_PRECISION)
+
+
+def staggered_cross_diff(expr, dims, deriv_order, fd_order, stagger):
+    """
+    Create a generic cross derivative for a given staggered function.
+
+    :param function: The symbol representing a function.
+    :param dims: Dimensions for which to take the derivative.
+    :param order: Discretisation order of the stencil to create.
+    """
+    first = staggered_diff(expr, deriv_order=deriv_order[0],
+                           fd_order=fd_order[0], dim=dims[0], stagger=stagger[0])
+    return staggered_diff(first, deriv_order=deriv_order[1],
+                          fd_order=fd_order[1], dim=dims[1], stagger=stagger[1])
+
+
+def generate_fd_functions(function):
+    """
+    Create all possible FD functions from the indices,
+    space_order (and time_order if time dependant)
+    """
+    dimensions = function.indices
+    space_fd_order = function.space_order
+    time_fd_order = function.time_order if function.is_TimeFunction else 0
+    is_staggered = any(s is not None for s in function.staggered)
+    deriv_function = staggered_diff if is_staggered else generic_derivative
+    c_deriv_function = staggered_cross_diff if is_staggered else generic_cross_derivative
+
+    side = dict()
+    for (d, s) in zip(dimensions, function.staggered):
+        if s == 0:
+            side[d] = left
+        elif s == 1:
+            side[d] = right
+        else:
+            side[d] = centered
+
+    derivatives = ()
+    done = ()
+
+    for d in dimensions:
+        # Dimension is treated, remove from list
+        done += (d,)
+        other_dims = tuple(i for i in dimensions if i not in done)
+        # Dimension name and corresponding FD order
+        dim_order = time_fd_order if d.is_Time else space_fd_order
+        name = 't' if d.is_Time else d.root.name
+        # All possible derivatives go up to the dimension FD order
+        for o in range(1, dim_order + 1):
+            deriv = partial(deriv_function, deriv_order=o, dim=d,
+                            fd_order=dim_order, stagger=side[d])
+            name_fd = 'd%s%d' % (name, o) if o > 1 else 'd%s' % name
+            desciption = 'derivative of order %d w.r.t dimension %s' % (o, d)
+
+            derivatives += ((deriv, name_fd, desciption),)
+            # Cross derivatives with the other dimension
+            # Skip already done dimensions a dxdy is the same as dydx
+            for d2 in other_dims:
+                dim_order2 = time_fd_order if d2.is_Time else space_fd_order
+                name2 = 't' if d2.is_Time else d2.root.name
```

Comment:
same

Reply:
I think what you're after here is:
``d.name if d2.is_Stepping else d.root.name``
(same before)